### PR TITLE
unix: strnlen is not available on Solaris 10

### PR DESCRIPTION
**Pull request by pekdon**
strnlen is not available, fall back to using memchr instead of adding compatibility code as this is the only use of strnlen in the tree

### CODE REVIEW

**Comment by bnoordhuis** [comment]
I appreciate that you took the time to open a pull request but, if I'm brutally honest, I'm rather lukewarm to the idea of merging this. `strnlen()` is specified in POSIX.2008, that's older than my oldest kid, and she's a teenager now.

I'm also kind of surprised anyone is still using Solaris 10. Can I ask where or what for?

**Comment by pekdon** [comment]
Would you oppose the idea of a compatibility implementation of strnlen on the side? Or is the whole idea of supporting old hardware/software a burden not worth the effort?

For me, this is a because I can and need the library compiling to get modern software up and running. Being a cross-platform library and using C89 style (and not lets say C99 at least) I thought there could be some room for this.

Anyway, if there's some way to get this in let me know what/how you want this changed and I'll do my best to cater for it. If not, thanks for being fast at replying. :)

**Comment by bnoordhuis** [comment]
> Would you oppose the idea of a compatibility implementation of strnlen on the side?

That'd be okay by me.

I also wouldn't have a problem with your change if it was isolated to `src/unix/sunos.c` but it's in a cross-platform file, it'd be tantamount to a "no strnlen() because $old_platform" rule for new development.

**Comment by pekdon** [comment]
Make sense. I would spontaneously put the code in a compat.c/compat.h, to allow for similar fixes without causing a file inflation in the source tree. Where would you want it to be placed?

**Comment by bnoordhuis** [comment]
Probably in `src/unix/sunos.c` with an ifdef-guarded prototype declaration in `src/unix/internal.h`. That way you don't have to tweak any build scripts.

**Comment by pekdon** [comment]
Something like that? Tried compiling on Solaris 10 and 11, seems to be OK.

File: src/unix/internal.h [hunk 1]
```diff
@@ -346,5 +346,11 @@ int uv__sendmmsg(int fd, struct uv__mmsghdr* mmsg, unsigned int vlen);
 #define HAVE_MMSG 0
 #endif
 
+#if defined(__sun)
+#if !defined(_POSIX_VERSION) || _POSIX_VERSION < 200809L
+size_t strnlen(const char *s, size_t maxlen);
```
Comment (bnoordhuis) [inline src/unix/internal.h:351]:
```suggestion
size_t strnlen(const char* s, size_t maxlen);
```

File: src/unix/sunos.c [hunk 2]
```diff
@@ -865,3 +865,14 @@ void uv_free_interface_addresses(uv_interface_address_t* addresses,
...
 }
+
+#if !defined(_POSIX_VERSION) || _POSIX_VERSION < 200809L
+size_t strnlen(const char *s, size_t maxlen) {
+  const char *end;
+  end = memchr(s, '\0', maxlen);
+  if (! end) {
+    return maxlen;
+  }
+  return end - s;
```
Comment (bnoordhuis) [inline src/unix/sunos.c:876]:
```suggestion
size_t strnlen(const char* s, size_t maxlen) {
  const char* end;
  end = memchr(s, '\0', maxlen);
  if (end == NULL)
    return maxlen;
  return end - s;
```

**Review by bnoordhuis — APPROVED** [review]
Some code style tweaks that I'll fix before landing but otherwise LGTM, thanks!

**Comment by pekdon** [comment]
> Some code style tweaks that I'll fix before landing but otherwise LGTM, thanks!

Great. Let me know if you want me to do something else related to this.

**Comment by bnoordhuis** [comment]
This is good to go, as far as I'm concerned. We usually leave pull requests open for a while and round them up before a release.

File: src/unix/sunos.c [hunk 3]
```diff
@@ -865,3 +865,13 @@ void uv_free_interface_addresses(uv_interface_address_t* addresses,
 
   uv__free(addresses);
 }
+
+#if !defined(_POSIX_VERSION) || _POSIX_VERSION < 200809L
```
Comment (cabelitos) [inline src/unix/sunos.c:869]: arent we missing the defined(__sun) in here?

Comment (pekdon) [reply src/unix/sunos.c:869]: A __sunos define in sunos.c shouldn't be required, right? :)

Comment (cabelitos) [reply src/unix/sunos.c:869]: My bad. Did not pay attention to the file name 😔. Really sorry.

**Review by vtjnash — APPROVED** [review]
@bnoordhuis Want to land this now? I would, but you mentioned wanting to fix style nits.

**Comment by vtjnash** [comment]
bump.

**Comment by vtjnash** [comment]
@bnoordhuis Can you note which style fixes you intended, since you had conditionally approved this for merging?

File: src/unix/sunos.c [hunk 4]
```diff
@@ -865,3 +865,13 @@ void uv_free_interface_addresses(uv_interface_address_t* addresses,
 
   uv__free(addresses);
 }
+
```
Comment (vtjnash) [inline src/unix/sunos.c:868]:
```suggestion


```

**Pull request merged**